analyzers/src/SonarAnalyzer.CSharp/Rules/CommentedOutCodeCodeFix.cs

```diff
@@ -0,0 +1,103 @@
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using Microsoft.CodeAnalysis.CodeActions;
+using Microsoft.CodeAnalysis.CodeFixes;
+using Trivias = System.Collections.Generic.Dictionary<Microsoft.CodeAnalysis.FileLinePositionSpan, Microsoft.CodeAnalysis.SyntaxTrivia>;
+
+namespace SonarAnalyzer.Rules.CSharp;
+
+[ExportCodeFixProvider(LanguageNames.CSharp)]
+public sealed class CommentedOutCodeCodeFix : SonarCodeFix
+{
+    public override ImmutableArray<string> FixableDiagnosticIds => ImmutableArray.Create(CommentedOutCode.DiagnosticId);
+
+    public override FixAllProvider GetFixAllProvider() => null;
+
+    protected override Task RegisterCodeFixesAsync(SyntaxNode root, CodeFixContext context)
+    {
+        var diagnostic = context.Diagnostics.First();
+
+        context.RegisterCodeFix(CodeAction.Create(
+            CommentedOutCode.MessageFormat,
+            c =>
+            {
+                var comment = diagnostic.Location.GetLineSpan();
+                var node = root.FindNode(diagnostic.Location.SourceSpan, getInnermostNodeForTie: true);
+
+                var newNode = node;
+
+                if (Remove(node.GetTrailingTrivia(), comment) is { } trailing)
+                {
+                    var token = node.ChildTokens().Last();
+                    ClearEmptiedLines(trailing, token, comment);
+                    RemoveWhitespace(trailing, token);
+                    newNode = node.WithTrailingTrivia(trailing.Values);
+                }
+                else if (Remove(node.GetLeadingTrivia(), comment) is { } leading)
+                {
+                    ClearEmptiedLines(leading, node.ChildTokens().First(), comment);
+                    newNode = node.WithLeadingTrivia(leading.Values);
+                }
+                else
+                {
+                    // this should never occur.
+                }
+
+                return Task.FromResult(context.Document.WithSyntaxRoot(root.ReplaceNode(node, newNode)));
+            }),
+            diagnostic);
+
+        return Task.CompletedTask;
+    }
+
+    private static Trivias Remove(SyntaxTriviaList trivias, FileLinePositionSpan comment)
+    {
+        var spans = trivias.ToDictionary(x => x.GetLocation().GetLineSpan(), x => x);
+        return spans.Remove(comment) ? spans : null;
+    }
+
+    private static void ClearEmptiedLines(Trivias triviaList, SyntaxToken token, FileLinePositionSpan comment)
+    {
+        if (!ShareLine(comment, token))
+        {
+            foreach (var shared in triviaList.Keys.Where(x => ShareLine(comment, x)))
```

Comment:
Here you are editing the collection you are enumerating inside the loop, which is an `InvalidOperation`.
Consider a local variable like:

```csharp
var redundantNewLines = triviaList.Keys.Where(x => ShareLine(comment, x)).ToList(); // ToList is necessary, to avoid deferred execution
```

...and enumerate that in the for-loop.

Reply:
The funny thing is, it does not. I'm not sure why, as the iterator of the KeyCollection does a version check. Anyhow, I've added `.ToArray()` (just a bit cheaper then `.ToList()`).

Reply:
I did...